File: GeneticAlgorithm/modules/path.py
class Path:
    """Class representing a Path.

    Attributes
        id:    The path id.

        cost:  The paths's cost.

        links: List of links the path uses.
    """

    def __init__(self, path_element):
        self.id = int(path_element.get('Id'))
        self.cost = float(path_element.get('Cost'))
        self.links = [int(link_element.get('Id')) for link_element
                      in path_element.findall('Link')]

    def __repr__(self):
        return 'Path(' + self.__str__() + ')'

    def __str__(self):
        return ('Path ID: {} Path Cost: {} Links used: {}'.format(self.id,
                                                                  self.cost,
                                                                  self.links))

File: KShortestPath/src/path.cpp
#include "path.hpp"

Path::Path(std::pair<linkCost_t, std::list<DefBoostGraph::link_t>> path,
           const std::map<DefBoostGraph::link_t,
                          DefLemonGraph::link_t>& m_blLinkMap) :
    pathCost (path.first)
{
  for (auto& link : path.second) // Populate list of links the path uses
    links.emplace_back (m_blLinkMap.at (link));
}

/**
 * @brief Overloads the Path::operator = to check if to Path objects are equal
 */
Path&
Path::operator= ( const Path& rhs )
{
  if (&rhs == this) return *this; // Check for self-assignment

  pathCost = rhs.pathCost;
  links = rhs.links;
  return *this;
}

/**
 * @brief Overloads the Path::operator < to compare to paths. Used for sorting
 */
bool
Path::operator< (const Path& other) const
{
  return pathCost < other.pathCost;
}

std::string Path::str (const LemonGraph &lemonGraph)
{
  std::stringstream ss;

  ss << "Path Cost: " << pathCost << "\n"
     << "Path: ";

  for (auto& link : links)
    ss << lemonGraph.GetLinkId (link) << " ";

  return ss.str();
}
